core/iwasm/libraries/wasi-nn/src/wasi_nn_onnx.cpp

```diff
@@ -0,0 +1,820 @@
+     execution_target target, graph *g)
+{
+    if (encoding != onnx) {
+        NN_ERR_PRINTF("Unsupported encoding: %d", encoding);
+        return invalid_encoding;
+    }
+
+    if (target != cpu) {
+        NN_ERR_PRINTF("Only CPU target is supported");
+        return unsupported_operation;
+    }
+
+    OnnxRuntimeContext *ctx = (OnnxRuntimeContext *)onnx_ctx;
+    std::lock_guard<std::mutex> lock(ctx->mutex);
+
+    int graph_index = -1;
+    for (int i = 0; i < MAX_GRAPHS; i++) {
+        if (!ctx->graphs[i].is_initialized) {
+            graph_index = i;
+            break;
+        }
+    }
+
+    if (graph_index == -1) {
+        NN_ERR_PRINTF("Maximum number of graphs reached");
+        return runtime_error;
+    }
+
+    if (builder->size == 0 || builder->buf == NULL) {
+        NN_ERR_PRINTF("No model data provided");
+        return invalid_argument;
+    }
+
+    NN_INFO_PRINTF("[ONNX Runtime] Loading model of size %zu bytes...",
+                   builder->buf[0].size);
+
+    if (builder->buf[0].size > 16) {
+        NN_INFO_PRINTF(
+            "Model header bytes: %02x %02x %02x %02x %02x %02x %02x %02x",
+            ((uint8_t *)builder->buf[0].buf)[0],
+            ((uint8_t *)builder->buf[0].buf)[1],
+            ((uint8_t *)builder->buf[0].buf)[2],
+            ((uint8_t *)builder->buf[0].buf)[3],
+            ((uint8_t *)builder->buf[0].buf)[4],
+            ((uint8_t *)builder->buf[0].buf)[5],
+            ((uint8_t *)builder->buf[0].buf)[6],
+            ((uint8_t *)builder->buf[0].buf)[7]);
+    }
+
+    OrtStatus *status = ctx->ort_api->CreateSessionFromArray(
+        ctx->env, builder->buf[0].buf, builder->buf[0].size,
+        ctx->session_options, &ctx->graphs[graph_index].session);
+
+    if (status != nullptr) {
+        const char *error_message = ctx->ort_api->GetErrorMessage(status);
+        wasi_nn_error err = convert_ort_error_to_wasi_nn_error(ctx, status);
+        NN_ERR_PRINTF("Failed to create ONNX Runtime session: %s",
+                      error_message);
+        ctx->ort_api->ReleaseStatus(status);
+        return err;
+    }
+
+    NN_INFO_PRINTF("ONNX Runtime session created successfully");
+
+    ctx->graphs[graph_index].is_initialized = true;
+    *g = graph_index;
+
+    NN_INFO_PRINTF("ONNX model loaded as graph %d", graph_index);
+    return success;
+}
+
+__attribute__((visibility("default"))) wasi_nn_error
+load_by_name(void *onnx_ctx, const char *name, uint32_t filename_len, graph *g)
+{
+    OnnxRuntimeContext *ctx = (OnnxRuntimeContext *)onnx_ctx;
+    std::lock_guard<std::mutex> lock(ctx->mutex);
+
+    int graph_index = -1;
+    for (int i = 0; i < MAX_GRAPHS; i++) {
+        if (!ctx->graphs[i].is_initialized) {
+            graph_index = i;
+            break;
+        }
+    }
+
+    if (graph_index == -1) {
+        NN_ERR_PRINTF("Maximum number of graphs reached");
+        return runtime_error;
+    }
+
```

Comment:
just my doubt, add a protector about `name`? like `name` is empty or NULL

Reply:
in wasi-nn.c, seems wasi_nn_load_by_name has guaranteed the name correctness by copyin_and_nul_terminate